src/Features/CSharp/Portable/Completion/CompletionProviders/AwaitCompletionProvider.cs

```diff
@@ -0,0 +1,119 @@
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.MakeMethodAsynchronous;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
+{
+    /// <summary>
+    /// A completion provider for offering <see langword="await"/> keyword.
+    /// This is implemented separately, not as a keyword recommender as it contains extra logic for making container method async.
+    /// </summary>
+    /// <remarks>
+    /// The container is made async if and only if the containing method is returning a Task-like type.
+    /// </remarks>
+    [ExportCompletionProvider(nameof(AwaitCompletionProvider), LanguageNames.CSharp)]
+    [ExtensionOrder(After = nameof(KeywordCompletionProvider))]
+    [Shared]
+    internal sealed class AwaitCompletionProvider : LSPCompletionProvider
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public AwaitCompletionProvider()
+        {
+        }
+
+        public override ImmutableHashSet<char> TriggerCharacters => CompletionUtilities.CommonTriggerCharactersWithArgumentList;
+
+        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            var document = context.Document;
+            var position = context.Position;
+            var cancellationToken = context.CancellationToken;
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var workspace = document.Project.Solution.Workspace;
+            var syntaxContext = CSharpSyntaxContext.CreateContext(workspace, semanticModel, position, cancellationToken);
+
+            if (!syntaxContext.IsAwaitKeywordContext(position))
+            {
+                return;
+            }
+
+            var method = syntaxContext.TargetToken.GetAncestor(node => node.IsAsyncSupportingFunctionSyntax());
+            if (method is not null && !method.GetModifiers().Any(SyntaxKind.AsyncKeyword))
+            {
+                var asyncService = document.GetRequiredLanguageService<IMakeMethodAsynchronousService>();
+                var symbol = semanticModel.GetDeclaredSymbol(method, cancellationToken) as IMethodSymbol;
+                symbol ??= semanticModel.GetSymbolInfo(method, cancellationToken).Symbol as IMethodSymbol;
+                if (symbol is not null && asyncService.IsAsyncReturnType(symbol.ReturnType, new KnownTaskTypes(semanticModel.Compilation)))
```

Comment:
interesting.  what's the thinking here?  if i'm adding 'await' it only makes sense if i make the container async.  Otherwise, i'm going to immediately error.  So i'd be somewhat ok with us unilaterally doing this without checking the return type of the symbol.

Reply:
@CyrusNajmabadi Currently, `async` is only added for Task-like return types. This is what I understood from the original issue.

Reply:
i feel like always adding is probablyt he right thing.  after all... if someone is typing 'await' it could only work if the containing member became async.  Now, it's more of an open question if we should *change* teh return type.  For now, let's just add 'async' unless there's a good reason not to.  (is there?)

Reply:
Alright, will make the change by the end of the week. I think I'll need to revert all the refactoring done to share code between this completion provider and the existing codefix providers.